### PR TITLE
A new ROM implementation of Rayleigh-Taylor instability problem

### PR DESCRIPTION
**Pull request by siuwuncheung**
The Rayleigh-Taylor Instability Problem can be run as follows:
`./laghos -p 7 -m data/rt2D.mesh -tf 4 -rs 1 -ok 4 -ot 3 -vis -pa`
The figures are taken at t = 4.0, and the formation of the ”mushroom cap” can be seen.

The following results are generated using the main directory which mirrors the most recent `master`.

Density at `t = 4.0`

master_density

Velocity at `t = 4.0`

master_velocity

Energy at `t = 4.0`

master_energy

### CODE REVIEW

**Comment by siuwuncheung** [comment]
All regression tests passed up to this point. 

Density at `t = 4.0`
density

Velocity at `t = 4.0`
velocity

Energy at `t = 4.0`
energy

**Comment by siuwuncheung** [comment]
After a change in boundary condition, regression tests are broken as expected, but with negligible error.

Density at `t = 4.0`
density

Velocity at `t = 4.0`
velocity

Energy at `t = 4.0`
energy

File: laghos.cpp [hunk 1]
```diff
@@ -605,7 +613,27 @@ int main(int argc, char *argv[])
    BlockVector S_old(S);
    long mem=0, mmax=0, msum=0;
    int checks = 0;
-
+   //   const double internal_energy = hydro.InternalEnergy(e_gf);
```
Comment (dylan-copeland) [inline laghos.cpp:616]: If this commented block of code should be kept for convenience, it would be better to use a `bool` to decide whether to do the output rather than commenting out.

File: laghos.cpp [hunk 2]
```diff
@@ -663,15 +691,24 @@ int main(int argc, char *argv[])
...
          {
             const double sqrt_norm = sqrt(norm);
+
             cout << std::fixed;
             cout << "step " << std::setw(5) << ti
                  << ",\tt = " << std::setw(5) << std::setprecision(4) << t
                  << ",\tdt = " << std::setw(5) << std::setprecision(6) << dt
                  << ",\t|e| = " << std::setprecision(10) << std::scientific
                  << sqrt_norm;
+            //  << ",\t|IE| = " << std::setprecision(10) << std::scientific
```
Comment (dylan-copeland) [inline laghos.cpp:706]: Again, use a condition rather than commenting out.

File: laghos.cpp (hunk 1 above)
Comment (siuwuncheung) [reply laghos.cpp:616]: These changes made to the main directory are actually merged from the `master` branch. I guess we should better not make any changes so that when we merge from the `master` in the future, it would not cause any conflicts.

File: laghos.cpp (hunk 2 above)
Comment (siuwuncheung) [reply laghos.cpp:706]: These changes made to the main directory are actually merged from the `master` branch. I guess we should better not make any changes so that when we merge from the `master` in the future, it would not cause any conflicts.

File: laghos.cpp (hunk 1 above)
Comment (dylan-copeland) [reply laghos.cpp:616]: Yes, I see now that it is in master, which is surprising. I agree to keep it, to avoid conflicts.

File: rom/laghos.cpp [hunk 3]
```diff
@@ -611,20 +614,23 @@ int main(int argc, char *argv[])
...
     {
         {
-            Array<int> ess_bdr(pmesh->bdr_attributes.Max()), tdofs1d;
+            Array<int> ess_bdr(pmesh->bdr_attributes.Max()), dofs_marker, dofs_list;
             for (int d = 0; d < pmesh->Dimension(); d++)
             {
                 // Attributes 1/2/3 correspond to fixed-x/y/z boundaries, i.e., we must
                 // enforce v_x/y/z = 0 for the velocity components.
-                ess_bdr = 0;
+                ess_bdr = 0;
```
Comment (dylan-copeland) [inline rom/laghos.cpp]: Extra space. Apparently `make style` does not eliminate this.

**Comment by dylan-copeland** [comment]
The test `./laghos -p 7 -m data/rt2D.mesh -tf 4 -rs 1 -ok 4 -ot 3 -vis -pa` works in parallel (e.g. `mpirun -n 4`) for the main version, but for the `rom/` version the timestep size goes to zero. @siuwuncheung have you looked into the parallel case?

**Comment by siuwuncheung** [comment]
> The test `./laghos -p 7 -m data/rt2D.mesh -tf 4 -rs 1 -ok 4 -ot 3 -vis -pa` works in parallel (e.g. `mpirun -n 4`) for the main version, but for the `rom/` version the timestep size goes to zero. @siuwuncheung have you looked into the parallel case?

I have been testing the serial runs only. I will look into the parallel runs.

**Comment by siuwuncheung** [comment]
> > The test `./laghos -p 7 -m data/rt2D.mesh -tf 4 -rs 1 -ok 4 -ot 3 -vis -pa` works in parallel (e.g. `mpirun -n 4`) for the main version, but for the `rom/` version the timestep size goes to zero. @siuwuncheung have you looked into the parallel case?
> 
> I have been testing the serial runs only. I will look into the parallel runs.

The parallel runs have been fixed.

**Review by dylan-copeland — APPROVED** [review]
Thanks @siuwuncheung I verified that the parallel run works. The failing Taylor-Green regression tests are acceptable to me, although the cause is unclear, as I verified that the errors and speed-up for the Taylor-Green test in our paper have not changed significantly.

**Review by chldkdtn — APPROVED** [review]

**Review by kevinhkhuynh — APPROVED** [review]
